### PR TITLE
Implement support for the Siglent SDL10x0 electronic loads.

### PR DESCRIPTION
**Pull request by NetSpida**
This adds very basic support for the SIGLENT SDL1020 and SDL1030 electronic loads (tested with SDL1020). Only Constant Current/Voltage/Power/Resistance are implemented yet, the advanced modes (LED, Battery, Program) are not yet implemented.

### CODE REVIEW

File: src/hardware/siglent-sdl10x0/api.c [hunk 1]
```diff
@@ -0,0 +1,461 @@
...
+	devc = g_malloc0(sizeof(struct dev_context));
+	sdi->priv = devc;
+
+	// modelname 'SDL1020X-E': 6th character indicates wattage: 2 => 200, 3 => 300
+	devc->maxpower = 200.0;
+	if (g_ascii_strncasecmp(sdi->model, "SDL1030", strlen("SDL1030")) == 0) {
+		devc->maxpower = 300.0;
+	}
+
+
```
Comment (biot) [inline src/hardware/siglent-sdl10x0/api.c:129]: Please remove extra blank lines, all through this patch.

File: src/hardware/siglent-sdl10x0/api.c [hunk 2]
```diff
@@ -0,0 +1,461 @@
...
+	SR_CONF_OVER_CURRENT_PROTECTION_ACTIVE | SR_CONF_GET,
+	SR_CONF_OVER_CURRENT_PROTECTION_THRESHOLD | SR_CONF_GET | SR_CONF_SET,
+};
+
+static const char *regulation[] = {
+	"CURRENT",
+	"VOLTAGE",
+	"POWER",
+	"RESISTANCE",
+	"LED"
```
Comment (biot) [inline src/hardware/siglent-sdl10x0/api.c:73]: What does "LED" refer to here? The rest of the driver seems to be unaware of this mode, other than the user can send it to the hardware.

File: src/hardware/siglent-sdl10x0/api.c [hunk 3]
```diff
@@ -0,0 +1,461 @@
...
+	ch = sr_channel_new(sdi, 0, SR_CHANNEL_ANALOG, TRUE, "1");
+	cg->channels = g_slist_append(cg->channels, ch);
+
+	sr_scpi_hw_info_free(hw_info);
+	hw_info = NULL;
+
+	devc = g_malloc0(sizeof(struct dev_context));
+	sdi->priv = devc;
+
+	// modelname 'SDL1020X-E': 6th character indicates wattage: 2 => 200, 3 => 300
```
Comment (biot) [inline src/hardware/siglent-sdl10x0/api.c:123]: /* */ style comments please, in other places too.

File: src/hardware/siglent-sdl10x0/api.c [hunk 4]
```diff
@@ -0,0 +1,461 @@
...
+	const struct sr_dev_inst *sdi, const struct sr_channel_group *cg)
+{
+	int ret;
+	float ival;
+	gboolean bval;
+	char *mode;
+
+	(void)sdi;
+	(void)data;
+	(void)cg;
```
Comment (biot) [inline src/hardware/siglent-sdl10x0/api.c:165]: These (void) statements are just in the stubs generated by `new-driver` to make the generated patch compile without warnings about unused variable. You're supposed to remove them as you fill out the stubs (other functions too).

File: src/hardware/siglent-sdl10x0/api.c [hunk 5]
```diff
@@ -0,0 +1,461 @@
...
+	.dev_list = std_dev_list,
+	.dev_clear = std_dev_clear,
+	.config_get = config_get,
+	.config_set = config_set,
+	.config_list = config_list,
+	.dev_open = dev_open,
+	.dev_close = dev_close,
+	.dev_acquisition_start = dev_acquisition_start,
+	.dev_acquisition_stop = dev_acquisition_stop,
+	.context = NULL,
```
Comment (biot) [inline src/hardware/siglent-sdl10x0/api.c:459]: No need to fill NULL entries.

**Review by biot — CHANGES_REQUESTED** [review]
- The original skeleton driver patch generated by `new-driver` is supposed to be in your patch set. That way we can review your changes easier when you fill in the non-skeleton parts. Just regenerate, commit it, and drop your code on top to get there.
- Split the libsigrok.h/hwdriver.c into their own separate (first) patch, those are infrastructure.

Other than the minor changes I requested, the driver looks really good. Are you planning on tackling the other modes this hardware can support?

File: src/hardware/siglent-sdl10x0/api.c (hunk 5 above)
Comment (NetSpida) [reply src/hardware/siglent-sdl10x0/api.c:452]: I found [a](https://github.com/sigrokproject/libsigrok/blob/0dfa850efb5553940982633ab96a8612860e87b4/src/hardware/korad-kaxxxxp/api.c#L546) [number](https://github.com/sigrokproject/libsigrok/blob/0dfa850efb5553940982633ab96a8612860e87b4/src/hardware/maynuo-m97/api.c#L428) [of](https://github.com/sigrokproject/libsigrok/blob/0dfa850efb5553940982633ab96a8612860e87b4/src/hardware/mastech-ms6514/api.c#L215) [examples](https://github.com/sigrokproject/libsigrok/blob/0dfa850efb5553940982633ab96a8612860e87b4/src/hardware/manson-hcs-3xxx/api.c#L351) where this was still in the driver, so I left it in for consistency. I can remove it if  you wish, though.

File: src/hardware/siglent-sdl10x0/protocol.c [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+	if (!devc) {
+		return TRUE;
+	}
+
+	if (revents & G_IO_IN) {
+		siglent_sdl10x0_receive_data(sdi);
+	} else {
+		sr_dev_acquisition_stop(sdi);
+		return FALSE;
+	}
```
[the remark this answers is not in the document]
Comment (NetSpida) [reply src/hardware/siglent-sdl10x0/protocol.c:172]: Thanks for the hint, that is of course unnecessary.

**Comment by NetSpida** [comment]
Thanks for the review.
I was not planning to add the other modes (yet), though I might do that eventually. Especially for the programming mode, which will run a sequence of the other modes (CC, CV, CR) with a provided configuration, the user interface should be discussed first.

**Review by biot — CHANGES_REQUESTED** [review]

File: src/hardware/siglent-sdl10x0/api.c (hunk 5 above)
Comment (biot) [reply src/hardware/siglent-sdl10x0/api.c:452]: Yeah you're right, and I see it's in the new-driver template as well. Never mind, it needs a patch to fix up all of them.

**Comment by biot** [comment]
> Especially for the programming mode, which will run a sequence of the other modes (CC, CV, CR) with a provided configuration, the user interface should be discussed first.

True, and it'll also take some comparing how other programmable loads/PSUs handle this as well. Perhaps @knarfS would like to weigh in, seeing how smuview is pretty much the reference client for this gear?

File: src/hwdriver.c [hunk 7]
```diff
@@ -252,6 +252,15 @@ static struct sr_key_info sr_key_info_config[] = {
...
+		"Over-power protection enabled", NULL},
+	{SR_CONF_OVER_POWER_PROTECTION_ACTIVE, SR_T_BOOL, "opp_active",
+		"Over-power protection active", NULL},
+	{SR_CONF_OVER_POWER_PROTECTION_THRESHOLD, SR_T_FLOAT, "opp_threshold",
+		"Over-power protection threshold", NULL},
+
+	{SR_CONF_RESISTANCE, SR_T_FLOAT, "resistance",
+		"Resistance", NULL},
 	ALL_ZERO
 };
```
[the remark this answers is not in the document]
Comment (NetSpida) [reply src/hwdriver.c:266]: Done, thanks for the review.

File: src/hardware/siglent-sdl10x0/api.c (hunk 3 above)
Comment (biot) [reply src/hardware/siglent-sdl10x0/api.c:122]: No seriously, /* */ style comments please, here and in two more places.

**Review by biot — CHANGES_REQUESTED** [review]

**Comment by biot** [comment]
Not sure if you're seeing these, but I keep asking about some // style comment fixes. You marked it as resolved and I think that makes my further comments invisible to you. I can't seem to "unresolve" a comment.

File: src/hardware/siglent-sdl10x0/api.c (hunk 3 above)
Comment (NetSpida) [reply src/hardware/siglent-sdl10x0/api.c]: Sorry, didn't see this

**Comment by NetSpida** [comment]
You were right, I didn't see the added comment that I marked resolved. Sorry. Should be fixed now.

**Comment by biot** [comment]
I'm almost embarrassed to ask, but could you also indent that first comment? Also, just squash that comment commit into the previous one. After that, this driver is ready to be merged. Thanks!

**Comment by NetSpida** [comment]
no problem at all. Thanks again for the review.

**Review by biot — APPROVED** [review]

**Comment by abraxa** [comment]
Thanks for your contribution, it was merged as https://sigrok.org/gitaction/libsigrok.git/7c9cf3a..cd468e3 .

**Pull request closed without merging**